### PR TITLE
retry clientCA post start hook on transient failures

### PR DESCRIPTION
**Pull request by deads2k**
@smarterclayton retries the poststarthook you saw failing.

Having looked through, it seems that I didn't kill the server on the failure.

### CODE REVIEW

**Comment by piosz** [comment]
@deads2k I'm not familiar with this code. Please find a better reviewer;)

File: pkg/master/client_ca_hook.go [hunk 1]
```diff
@@ -38,25 +41,38 @@ type ClientCARegistrationHook struct {
...
 	}
 
-	h.writeClientCAs(client)
 	return nil
+
 }
 
 // writeClientCAs is here for unit testing with a fake client
-func (h ClientCARegistrationHook) writeClientCAs(client coreclient.CoreInterface) {
+func (h ClientCARegistrationHook) writeClientCAs(client coreclient.CoreInterface) (bool, error) {
```
Comment (sttts) [inline pkg/master/client_ca_hook.go]: godoc for the bool

File: pkg/master/client_ca_hook.go [hunk 2]
```diff
@@ -67,34 +83,34 @@ func (h ClientCARegistrationHook) writeClientCAs(client coreclient.CoreInterface
...
 
+	// write errors may work next time if we retry, so queue for retry
 	if err := writeConfigMap(client, "extension-apiserver-authentication", data); err != nil {
 		utilruntime.HandleError(err)
+		return false, nil
 	}
 
-	return
+	return true, nil
+
```
Comment (sttts) [inline pkg/master/client_ca_hook.go]: no newline

File: pkg/master/client_ca_hook.go (hunk 1 above)
Comment (sttts) [inline pkg/master/client_ca_hook.go]: Just call it tryToWriteClientCAs. Then it's more obvious.

File: pkg/master/client_ca_hook.go [hunk 3]
```diff
@@ -38,25 +41,38 @@ type ClientCARegistrationHook struct {
...
 	if len(h.ClientCA) == 0 && len(h.RequestHeaderCA) == 0 {
 		return nil
 	}
 
-	client, err := coreclient.NewForConfig(hookContext.LoopbackClientConfig)
+	// intializing CAs is important so that aggregated API servers can come up with "normal" config.
+	// We've seen lagging etcd before, so we want to retry this a few times before we decide to crashloop
+	// the API server on it.
+	err := wait.Poll(1*time.Second, 30*time.Second, func() (done bool, err error) {
+		client, err := coreclient.NewForConfig(hookContext.LoopbackClientConfig)
```
Comment (sttts) [inline pkg/master/client_ca_hook.go]: why retry the NewForConfig? If it important, comment please.

Comment (deads2k) [inline pkg/master/client_ca_hook.go]: > why retry the NewForConfig? If it important, comment please.

If memory serves, a new client sometimes makes a discovery request.  I remember hitting in the RBAC post start hook.

**Comment by deads2k** [comment]
comments addressed.

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**